Offloading documentation into the official documentation (#594)
buildkite/agent-stack-k8s: DEVELOPMENT.md

3. The name of your Buildkite organization (slug) and your target Buildkite Cluster UUID.
4. Depending on test cases, you may also need SSH keys, please keep reading.
5. Your shell environment will need CLI write access to a Kubernetes cluster such as the one provided by https://orbstack.dev/.

#### Using environment variables

It's generally convenient to supply the API token, your Buildkite organization name, and cluster UUID as environment variables. This can be done using an `.envrc` file loaded by using [direnv](https://direnv.net/).

```bash
export BUILDKITE_TOKEN="bkua_**************"
export ORG="your-cool-org-slug"
export CLUSTER_UUID="UUID-UUID-UUID-UUID"
```

## Running locally

To run the controller locally, with the environment variables, run the following example. Note that in this example the queue is overridden to ensure jobs from the default queue, which is "", are picked up by the Buildkite agent.

```bash
just run --org $ORG --buildkite-token $BUILDKITE_TOKEN --debug --tags 'queue=,os=linux'
```

## Testing locally

Before you start, check which Kubernetes cluster configuration you are using by default:

> ```bash

Reviewer: There were a series of these code blocks missing the `bash` syntax styler. I hope you don't mind me adding these to the code blocks that were missing them as my IDE's linter tool kept pointing them out to me.

Anyway, I guess this makes them more consistent with the other code blocks on this page.

```

> During test run, the test suites:

Reviewer 2: (Comment about the existing document) May be worth being more specific? Or more specific about the abstract description:

```suggestion
During a test run, each integration test generally performs these steps:
```

1. Create ephemeral pipelines and queues for a given [Buildkite Agent Cluster](https://buildkite.com/docs/clusters/overview).

> 2. Run executor, which will monitor jobs from the target queue in target Buildkite Cluster and start new Jobs in a Kubernetes cluster.

Reviewer 2: (Comment about the existing document) This should probably refer to the controller, since "executor" is not the name of the stack component.

There's also a missing step or two.

```suggestion
2. Runs the controller, which will monitor jobs from the (just-created) queue in the Buildkite Cluster and start new Jobs in the Kubernetes cluster.
3. Starts a build of the pipeline on Buildkite, which causes Buildkite jobs to become available.
4. Polls Buildkite while waiting for the expected outcome, which may include build success, build failure, and the presence or absence of certain log messages.
```

> 3. Clean up those ephemeral objects in the end.

Reviewer 2: (Comment about the existing document) Not sure how this feels stylistically, but worth calling out which ephemeral objects are cleaned up? (The Kubernetes jobs and pods are also ephemeral objects, but not ones mentioned in step 1.)

```suggestion
5. Cleans up those ephemeral objects (pipelines and queues).
```

To run integration test locally, we recommend you to run individual tests. For example,

```bash
just test -run TestWalkingSkeleton
```

### Integration test requirements

In addition to the usual requirements, the integration tests make use of a GraphQL-enabled Buildkite API token in order to create pipelines, start builds, read logs, and clean up afterwards.

### Setup

Any Buildkite user who has an access to a Kubernetes cluster should be able to run our integration test.

To get the integration test running locally, you will need:

> 1. A valid Buildkite API token.

Reviewer 2: ```suggestion
1. A valid Buildkite API token with GraphQL enabled.
```

> 2. A valid Buildkite Agent Token in your target Buildkite Cluster. The agent token needs to be installed in your Kubernetes

Reviewer 2: (Comment about the existing document) Since the instructions for installing the token are further down, I don't think this sentence ("The agent token needs to be installed ...") adds value.

Author: Done, thanks, @DrJosh9000!

## Cleanup

In general, for successful tests, pipelines will be deleted automatically. However, for unsuccessful tests, then will remain after then end of the test job to allow you to debug them.

To do clean them up, run:

```bash
just cleanup-orphans
```

> For this to work, you will need a "classic" GraphQL-enabled token as well as:

Reviewer 2: ```suggestion
For this to work, you will need a Buildkite API token with GraphQL enabled and these REST API scopes also enabled:
```

or something

- `read_artifacts`
- `write_pipelines`

This is usually enough, but there is another situation where the cluster could be clogged with Kubernetes jobs.
To clean these out, you should run the following in a Kubernetes context in the namespace containing the controller used to run the CI pipeline.

```bash
kubectl get -o jsonpath='{.items[*].metadata.name}' jobs | xargs -L1 kubectl delete job
```

### CI ❤️  integration test

> At the time of writing, the CI pipeline run in an EKS cluster, `agent-stack-k8s-ci` in the `buildkite-agent` AWS account.

Reviewer 2: ```suggestion
At the time of writing, the CI pipeline run in an EKS cluster, `agent-stack-k8s-ci` in the `buildkite-dist` AWS account.
```

The `config` key contains configuration passed directly to the binary, and so supports all the keys documented in [the example](examples/config.yaml).

> ## Release

Reviewer 2: (Comment about the existing document)

```suggestion
## Release procedure
```
